# Investigating Data of Hackers News Posts

### Hacker News Posts
Hacker News, a popular site where technology related stories (or 'posts') are voted and commented upon. 

The two types of posts we'll explore
- Ask HN 
- Show HN.

### Work flow of Hacker News Posts
Users submit Ask HN posts to ask the Hacker News community a specific question, such as "What is the best online course you've ever taken?" Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting.

### Objective of this notebook
We'll specifically compare these two types of posts to determine the following:

- Do Ask HN or Show HN receive more comments on average?
- Do posts created at a certain time receive more comments on average?

*It should be noted that the data set we're working with was reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions.*

In [1]:
# using csv module to read the hacker-news.csv
import csv

file_opened = open('hacker_news.csv')
read_file = csv.reader(file_opened)

# converting to list
hacker_news = list(read_file)

# extract header
header = hacker_news[0]

# view header
print(header)

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


We can see above that the data set contains the title of the posts, the number of comments for each post, the author of each post and the date the post was created. 

In [2]:
# extract data
hacker_news = hacker_news[1:]

# view first five rows of data
hacker_news[:5]

[['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01'],
 ['10301696',
  'Note by Note: The Making of Steinway L1037 (2007)',
  'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0',
  '8',
  '2',
  'walterbell',
  '9/30/2015 4:12']]

## Identifying and Separating Ask HN and Show HN Posts
First, we'll identify posts that begin with either Ask HN or Show HN and separate the data for those two types of posts into different lists. Separating the data makes it easier to analyze in the following steps.

In [3]:
# Identify posts that begin with either `Ask HN` or `Show HN` and separate the data into different lists.
ask_posts = []
show_posts =[]
other_posts = []

for row in hacker_news:
    title_name = row[1]
    if title_name.lower().startswith("ask hn"):
        ask_posts.append(row)
    elif title_name.lower().startswith("show hn"):
        show_posts.append(row)
    else:
        other_posts.append(row)
        
print("Number of Ask HN posts are {} ".format(len(ask_posts)))
print("Number of Show HN posts are {} ".format(len(show_posts)))
print("Number of Other posts are {} ".format(len(other_posts)))

Number of Ask HN posts are 1744 
Number of Show HN posts are 1162 
Number of Other posts are 17194 


In [4]:
# display first five ask hn posts
ask_posts[:5]

[['12296411',
  'Ask HN: How to improve my personal website?',
  '',
  '2',
  '6',
  'ahmedbaracat',
  '8/16/2016 9:55'],
 ['10610020',
  'Ask HN: Am I the only one outraged by Twitter shutting down share counts?',
  '',
  '28',
  '29',
  'tkfx',
  '11/22/2015 13:43'],
 ['11610310',
  'Ask HN: Aby recent changes to CSS that broke mobile?',
  '',
  '1',
  '1',
  'polskibus',
  '5/2/2016 10:14'],
 ['12210105',
  'Ask HN: Looking for Employee #3 How do I do it?',
  '',
  '1',
  '3',
  'sph130',
  '8/2/2016 14:20'],
 ['10394168',
  'Ask HN: Someone offered to buy my browser extension from me. What now?',
  '',
  '28',
  '17',
  'roykolak',
  '10/15/2015 16:38']]

In [5]:
# display first five show hn posts
show_posts[:5]

[['10627194',
  'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform',
  'https://iot.seeed.cc',
  '26',
  '22',
  'kfihihc',
  '11/25/2015 14:03'],
 ['10646440',
  'Show HN: Something pointless I made',
  'http://dn.ht/picklecat/',
  '747',
  '102',
  'dhotson',
  '11/29/2015 22:46'],
 ['11590768',
  'Show HN: Shanhu.io, a programming playground powered by e8vm',
  'https://shanhu.io',
  '1',
  '1',
  'h8liu',
  '4/28/2016 18:05'],
 ['12178806',
  'Show HN: Webscope  Easy way for web developers to communicate with Clients',
  'http://webscopeapp.com',
  '3',
  '3',
  'fastbrick',
  '7/28/2016 7:11'],
 ['10872799',
  'Show HN: GeoScreenshot  Easily test Geo-IP based web pages',
  'https://www.geoscreenshot.com/',
  '1',
  '9',
  'kpsychwave',
  '1/9/2016 20:45']]

## Objective 1: Do Ask HN or Show HN receive more comments on average?

Now that we separated ask posts and show posts into different lists, we'll calculate the average number of comments each type of post receives.

In [6]:
# calculate total comments on ask posts
ask_posts_total_comments = 0

# total number of ask posts
n_ask_posts = len(ask_posts)

# adding up all the comments
for row in ask_posts:
    n_comments = int(row[4])
    ask_posts_total_comments += n_comments

# calculating the average of comments made on asks posts
ask_post_average = round(ask_posts_total_comments/n_ask_posts,6)

print("The sum of ask posts is {}".format(ask_posts_total_comments))
print("The average of ask posts is {}".format(ask_post_average))

The sum of ask posts is 24483
The average of ask posts is 14.038417


In [7]:
# calculate total comments on show posts
show_posts_total_comments = 0

# total number of show posts
n_show_posts = len(show_posts)

# adding up all the comments
for row in show_posts:
    n_comments = int(row[4])
    show_posts_total_comments += n_comments

# calculating the average of comments made on asks posts
show_posts_average = round(show_posts_total_comments/n_show_posts,6)

print("The sum of ask posts is {}".format(show_posts_total_comments))
print("The average of ask posts is {}".format(show_posts_average))

The sum of ask posts is 11988
The average of ask posts is 10.316695


## Conclusion
On average, ask posts in our sample receive approximately **14 comments average**, whereas show posts receive approximately **10 comments average**. 

## Objective 2: Do posts created at a certain time receive more comments on average?

Next, we'll determine if there was a great traffic at a certain time. First, we'll find the amount of ask posts created during each hour of day, along with the number of comments those posts received. Then, we'll calculate the average amount of comments ask posts created at each hour of the day receive.

In [13]:
# for manipulating and parsing dates
import datetime as dt

# creating two dictionaries
count_by_hour = {}
comments_by_hour = {}

# specifying date format recorded in dataset
date_format = "%m/%d/%Y %H:%M"

for row in ask_posts:
    date = row[6]
    n_comments = int(row[4])
    #striping hours out of data
    time = dt.datetime.strptime(date, date_format).strftime("%H")
    
    if time in count_by_hour:
        count_by_hour[time]+=1
        comments_by_hour[time] += n_comments
    else:
        count_by_hour[time]=1
        comments_by_hour[time]=1

In [14]:
print(count_by_hour)

{'22': 71, '20': 80, '23': 68, '21': 109, '10': 59, '18': 109, '17': 100, '01': 60, '15': 116, '03': 54, '07': 34, '04': 47, '11': 58, '14': 107, '19': 110, '00': 55, '13': 85, '16': 108, '02': 58, '05': 46, '06': 44, '09': 45, '12': 73, '08': 48}


In [12]:
print(comments_by_hour)

{'22': 478, '20': 1721, '23': 543, '21': 1742, '10': 793, '18': 1438, '17': 1146, '01': 651, '15': 4477, '03': 421, '07': 266, '04': 335, '11': 640, '14': 1414, '19': 1186, '00': 438, '13': 1225, '16': 1798, '02': 1379, '05': 436, '06': 397, '09': 246, '12': 684, '08': 488}


In [16]:
# Calculate the average amount of comments `Ask HN` posts created at each hour of the day receive.
avg_by_hour = []

for hr in comments_by_hour:
    avg_by_hour.append([hr, comments_by_hour[hr] / count_by_hour[hr]])

avg_by_hour

[['22', 6.732394366197183],
 ['20', 21.5125],
 ['23', 7.985294117647059],
 ['21', 15.98165137614679],
 ['10', 13.440677966101696],
 ['18', 13.192660550458715],
 ['17', 11.46],
 ['01', 10.85],
 ['15', 38.5948275862069],
 ['03', 7.796296296296297],
 ['07', 7.823529411764706],
 ['04', 7.127659574468085],
 ['11', 11.03448275862069],
 ['14', 13.214953271028037],
 ['19', 10.781818181818181],
 ['00', 7.963636363636364],
 ['13', 14.411764705882353],
 ['16', 16.64814814814815],
 ['02', 23.775862068965516],
 ['05', 9.478260869565217],
 ['06', 9.022727272727273],
 ['09', 5.466666666666667],
 ['12', 9.36986301369863],
 ['08', 10.166666666666666]]

In [17]:
swap_avg_by_hour = []

for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
    
print(swap_avg_by_hour)

sorted_swap = sorted(swap_avg_by_hour, reverse=True)

sorted_swap

[[6.732394366197183, '22'], [21.5125, '20'], [7.985294117647059, '23'], [15.98165137614679, '21'], [13.440677966101696, '10'], [13.192660550458715, '18'], [11.46, '17'], [10.85, '01'], [38.5948275862069, '15'], [7.796296296296297, '03'], [7.823529411764706, '07'], [7.127659574468085, '04'], [11.03448275862069, '11'], [13.214953271028037, '14'], [10.781818181818181, '19'], [7.963636363636364, '00'], [14.411764705882353, '13'], [16.64814814814815, '16'], [23.775862068965516, '02'], [9.478260869565217, '05'], [9.022727272727273, '06'], [5.466666666666667, '09'], [9.36986301369863, '12'], [10.166666666666666, '08']]


[[38.5948275862069, '15'],
 [23.775862068965516, '02'],
 [21.5125, '20'],
 [16.64814814814815, '16'],
 [15.98165137614679, '21'],
 [14.411764705882353, '13'],
 [13.440677966101696, '10'],
 [13.214953271028037, '14'],
 [13.192660550458715, '18'],
 [11.46, '17'],
 [11.03448275862069, '11'],
 [10.85, '01'],
 [10.781818181818181, '19'],
 [10.166666666666666, '08'],
 [9.478260869565217, '05'],
 [9.36986301369863, '12'],
 [9.022727272727273, '06'],
 [7.985294117647059, '23'],
 [7.963636363636364, '00'],
 [7.823529411764706, '07'],
 [7.796296296296297, '03'],
 [7.127659574468085, '04'],
 [6.732394366197183, '22'],
 [5.466666666666667, '09']]

In [18]:
# Sort the values and print the the 5 hours with the highest average comments.

print("Top 5 Hours for 'Ask HN' Comments")
for avg, hr in sorted_swap[:5]:
    print(
        "{}: {:.2f} average comments per post".format(
            dt.datetime.strptime(hr, "%H").strftime("%H:%M"),avg
        )
    )

Top 5 Hours for 'Ask HN' Comments
15:00: 38.59 average comments per post
02:00: 23.78 average comments per post
20:00: 21.51 average comments per post
16:00: 16.65 average comments per post
21:00: 15.98 average comments per post


In [19]:
# creating two dictionaries
count_by_hour = {}
comments_by_hour = {}

# specifying date format recorded in dataset
date_format = "%m/%d/%Y %H:%M"

for row in show_posts:
    date = row[6]
    n_comments = int(row[4])
    #striping hours out of data
    time = dt.datetime.strptime(date, date_format).strftime("%H")
    
    if time in count_by_hour:
        count_by_hour[time]+=1
        comments_by_hour[time] += n_comments
    else:
        count_by_hour[time]=1
        comments_by_hour[time]=1

In [20]:
print(count_by_hour)

{'22': 46, '20': 60, '23': 36, '21': 47, '10': 36, '18': 61, '17': 93, '01': 28, '15': 78, '03': 27, '07': 26, '04': 26, '11': 44, '14': 86, '19': 55, '00': 31, '13': 99, '16': 93, '02': 30, '05': 19, '06': 16, '09': 30, '12': 61, '08': 34}


In [21]:
print(comments_by_hour)

{'22': 469, '20': 604, '23': 447, '21': 272, '10': 297, '18': 962, '17': 910, '01': 242, '15': 629, '03': 285, '07': 297, '04': 246, '11': 490, '14': 1135, '19': 539, '00': 487, '13': 946, '16': 1084, '02': 127, '05': 56, '06': 141, '09': 291, '12': 720, '08': 164}


In [22]:
# Calculate the average amount of comments `SHOW HN` posts created at each hour of the day receive.
avg_by_hour = []

for hr in comments_by_hour:
    avg_by_hour.append([hr, round(comments_by_hour[hr] / count_by_hour[hr]]))

avg_by_hour

[['22', 10.195652173913043],
 ['20', 10.066666666666666],
 ['23', 12.416666666666666],
 ['21', 5.787234042553192],
 ['10', 8.25],
 ['18', 15.770491803278688],
 ['17', 9.78494623655914],
 ['01', 8.642857142857142],
 ['15', 8.064102564102564],
 ['03', 10.555555555555555],
 ['07', 11.423076923076923],
 ['04', 9.461538461538462],
 ['11', 11.136363636363637],
 ['14', 13.19767441860465],
 ['19', 9.8],
 ['00', 15.709677419354838],
 ['13', 9.555555555555555],
 ['16', 11.655913978494624],
 ['02', 4.233333333333333],
 ['05', 2.9473684210526314],
 ['06', 8.8125],
 ['09', 9.7],
 ['12', 11.80327868852459],
 ['08', 4.823529411764706]]

In [24]:
swap_avg_by_hour = []

for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
    
print(swap_avg_by_hour)

sorted_swap = sorted(swap_avg_by_hour, reverse=True)

sorted_swap

[[10.195652173913043, '22'], [10.066666666666666, '20'], [12.416666666666666, '23'], [5.787234042553192, '21'], [8.25, '10'], [15.770491803278688, '18'], [9.78494623655914, '17'], [8.642857142857142, '01'], [8.064102564102564, '15'], [10.555555555555555, '03'], [11.423076923076923, '07'], [9.461538461538462, '04'], [11.136363636363637, '11'], [13.19767441860465, '14'], [9.8, '19'], [15.709677419354838, '00'], [9.555555555555555, '13'], [11.655913978494624, '16'], [4.233333333333333, '02'], [2.9473684210526314, '05'], [8.8125, '06'], [9.7, '09'], [11.80327868852459, '12'], [4.823529411764706, '08']]


[[15.770491803278688, '18'],
 [15.709677419354838, '00'],
 [13.19767441860465, '14'],
 [12.416666666666666, '23'],
 [11.80327868852459, '12'],
 [11.655913978494624, '16'],
 [11.423076923076923, '07'],
 [11.136363636363637, '11'],
 [10.555555555555555, '03'],
 [10.195652173913043, '22'],
 [10.066666666666666, '20'],
 [9.8, '19'],
 [9.78494623655914, '17'],
 [9.7, '09'],
 [9.555555555555555, '13'],
 [9.461538461538462, '04'],
 [8.8125, '06'],
 [8.642857142857142, '01'],
 [8.25, '10'],
 [8.064102564102564, '15'],
 [5.787234042553192, '21'],
 [4.823529411764706, '08'],
 [4.233333333333333, '02'],
 [2.9473684210526314, '05']]

In [25]:
# Sort the values and print the the 5 hours with the highest average comments.

print("Top 5 Hours for 'Show HN' Comments")
for avg, hr in sorted_swap[:5]:
    print(
        "{}: {:.2f} average comments per post".format(
            dt.datetime.strptime(hr, "%H").strftime("%H:%M"),avg
        )
    )

Top 5 Hours for 'Show HN' Comments
18:00: 15.77 average comments per post
00:00: 15.71 average comments per post
14:00: 13.20 average comments per post
23:00: 12.42 average comments per post
12:00: 11.80 average comments per post


## Conclusion
In this project, we have analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. 
Based on our analysis, ASK HN posts has the most comments with an average of 14 comments and the time at which most comments were done is 18:00 with 38.59